Add 'spearman' correlation method for `dataframe.corr`

dominicshanshan opened the pull request:
<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->
Closes #6804

Adds 'spearman' correlation method for `dataframe.corr`

GPUtester commented:
Can one of the admins verify this patch?

isVoid commented:
ok to test

isVoid reviewed (changes requested):
Thanks for contributing! Can you add unit tests for this method to match pandas output?

brandon-b-miller reviewed (changes requested):
Thanks very much for contributing this! Can you make it so that `method` is a kwarg that defaults to `pearson` so that we match the pandas API here? https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.corr.html

beckernick commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -6931,12 +6931,22 @@ def cov(self, **kwargs):
-        corr = cupy.corrcoef(self.values, rowvar=False)
-        df = DataFrame(cupy.asfortranarray(corr)).set_index(self.columns)
-        df.columns = self.columns
+    def corr(self, method):
```
It would be nice to include spearman correlation on the Series object, since it's being added to the DataFrame.

dominicshanshan replied:
will do, based on pandas  https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.corr.html

also will do unittest on this

dominicshanshan commented:
> Thanks for contributing! Can you add unit tests for this method to match pandas output?

here is the unit test script for DataFrame output, will added series calculation as identical as Series.corr in pandas API in coming days + unit test with Series.


[unittest_spearman.zip](https://github.com/rapidsai/cudf/files/6028130/unittest_spearman.zip)

isVoid commented:
@dominicshanshan The unitest looks great! In cudf, we utilize [pytest](https://docs.pytest.org/en/stable/) to organize our unit test process. For existing `corr` function, the coverage is defined in:
https://github.com/rapidsai/cudf/blob/c929ba1fe85c152d6e8b4c868cd36f0802dafa51/python/cudf/cudf/tests/test_stats.py#L396-L402

It's likely you can combine the `test_spearman` and `test_pearson` into `test_df_corr` function by parameterizing the `method` argument. An example of parameterization is at: 
https://github.com/rapidsai/cudf/blob/c929ba1fe85c152d6e8b4c868cd36f0802dafa51/python/cudf/cudf/tests/test_stats.py#L72-L73
where `test_series_std` will run 3 times with `ddof` parameterized as 0, 1 and 2.

It's also likely you will need https://github.com/rapidsai/cudf/blob/2234554c6cfe2941a30401bed907e409bf4bf56c/python/cudf/cudf/tests/utils.py#L69
to compare the result. You can specify `check_exact=False` for float equality.

dominicshanshan commented:
> **to compare the result. You can specify `check_exact=False` for float equality.**

I tested, it seems if your test datatype is float, it will use `assert_almost_equal ` directly, else will use `assert_equal`

isVoid commented:
Awesome, and please rebase this PR to `branch-0.19`.

dominicshanshan commented:
> Awesome, and please change the base of this PR to `branch-0.19`.


please find pytest version of unit test for spearman correlation function, I will also finish this feature in Series and pytest, will change to 0.19 brench  
[pytest_spearman_corr.zip](https://github.com/rapidsai/cudf/files/6040837/pytest_spearman_corr.zip)

isVoid commented:
Test looks great! Can you push the test code to this PR?

dominicshanshan commented:
> Test looks great! Can you push the test code to this PR?

Do you mean the pytest code?

isVoid commented on python/cudf/cudf/core/series.py:
```diff
@@ -3639,19 +3639,31 @@ def corr(self, other, method="pearson", min_periods=None):
+            
+            lhs = self.nans_to_nulls().dropna()
+            rhs = other.nans_to_nulls().dropna()
+            lhs, rhs = _align_indices([lhs, rhs], how="inner")
```
Can we consolidate this redundant line to just before the return statement?

dominicshanshan replied:
done

isVoid commented on python/cudf/cudf/core/series.py:
```diff
@@ -3639,19 +3639,31 @@ def corr(self, other, method="pearson", min_periods=None):
+            
+            lhs = self.nans_to_nulls().dropna().rank()
+            rhs = other.nans_to_nulls().dropna().rank()
+            lhs, rhs = _align_indices([lhs, rhs], how="inner")
```
Same as above.

dominicshanshan replied:
> lgtm, aside one small suggestion. Please also address @brandon-b-miller 's review above. Good job!

done

isVoid reviewed (approved):
lgtm, aside one small suggestion. Please also address @brandon-b-miller 's review above. Good job!

dominicshanshan commented:
> lgtm, aside one small suggestion. Please also address @brandon-b-miller 's review above. Good job!

done, @brandon-b-miller, please check and review

isVoid commented on python/cudf/cudf/core/series.py:
```diff
@@ -3630,7 +3630,7 @@ def cov(self, other, min_periods=None):
         return lhs._column.cov(rhs._column)
 
-    def corr(self, other, method="pearson", min_periods=None):
+    def corr(self, other, method, min_periods=None):
```
```suggestion
    def corr(self, other, method='pearson', min_periods=None):
```
according to https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.corr.html

isVoid reviewed (commented):
@brandon-b-miller can you give another look since you also requested change?

dominicshanshan commented:
Are we closed this PR?

beckernick commented:
@dominicshanshan we'd still like to support this feature (the "inactive-30d" message is automated). It looks like this pull request is almost ready to go based on the reviews/approvals/suggestions. Are you interested in finishing this pull request or would you prefer someone else jump in?

dominicshanshan commented:
> @dominicshanshan we'd still like to support this feature (the "inactive-30d" message is automated). It looks like this pull request is almost ready to go based on the reviews/approvals/suggestions. Are you interested in finishing this pull request or would you prefer someone else jump in?

no problem, I will finish this PR, let me know how I can help, as always :)

vyasr commented:
@dominicshanshan could you merge in the latest branch (now `branch-0.20`) and resolve conflicts? It looks like this is just waiting on an approval, so I can follow-up on that end once you get this merged up and have tests passing. Thanks!

isVoid commented:
ditto @vyasr, and please rebase the PR to `branch-0.20`.

brandon-b-miller commented:
This LGTM - just one quick comment about matching the pandas API by defaulting to `method='pearson'` and then this is ready to go.

vyasr commented:
@dominicshanshan hope you don't mind, I went ahead and took care of the merge and made a few other minor edits to try and get this merged soon.

vyasr commented:
rerun tests

vyasr commented:
rerun tests

ajschmidt8 commented:
ok to test

beckernick reviewed (approved):
LGTM

EDIT: Failing tests on arrays of different sizes. Presumably pandas is doing something to handle this scenario as the data would not be pairwise matched (interpolation?)

vyasr commented:
@dominicshanshan would you be able to look into that case? I think that's the last hurdle, then this can be merged.

dominicshanshan commented:
> LGTM
> 
> EDIT: Failing tests on arrays of different sizes. We should either throw or handle this. Presumably pandas is doing something to handle this scenario as the data would not be pairwise matched (interpolation?)

DataFrame can go with green light, Series is tricky need more time to test vs. pandas series, this week and next week is fully booked, will go back with test result and discuss with @beckernick if is necessary

harrism commented:
@dominicshanshan any plans to resume this?

dominicshanshan commented:
> @dominicshanshan any plans to resume this?

Sure! @harrism if you have the deadline or schedule for this fix please share with me, or let me do part of this fixing. Need to rearrange my time schedule for this

isVoid commented:
@dominicshanshan cuDF RAPIDS's release schedule as laid out here: https://docs.rapids.ai/maintainers. As of now, this feature is planned for v21.10 so please look out for the `Development(cuDF/rmm)` row on the 21.10 board.

dominicshanshan commented:
> @dominicshanshan cuDF follows RAPIDS's release schedule as laid out here: https://docs.rapids.ai/maintainers. As of now, this feature is planned for v21.10 so please look out for the `Development(cuDF/rmm)` end date on the 21.10 board.

@isVoid  Thanks for the info, I will schedule this feature testing in August

dominicshanshan commented:
> > @dominicshanshan cuDF follows RAPIDS's release schedule as laid out here: https://docs.rapids.ai/maintainers. As of now, this feature is planned for v21.10 so please look out for the `Development(cuDF/rmm)` end date on the 21.10 board.
> 
> @isVoid Thanks for the info, I will schedule this feature testing in August

@isVoid, can we postpone & plan this feature to v21.12? August is booked with other unexpected stuff, not sure I can rush out and finish in September, seems v21.12 timeline is suitable for me

dominicshanshan commented:
> > > @dominicshanshan cuDF follows RAPIDS's release schedule as laid out here: https://docs.rapids.ai/maintainers. As of now, this feature is planned for v21.10 so please look out for the `Development(cuDF/rmm)` end date on the 21.10 board.
> > 
> > 
> > @isVoid Thanks for the info, I will schedule this feature testing in August
> 
> @isVoid, can we postpone & plan this feature to v21.12? August is booked with other unexpected stuff, not sure I can rush out and finish in September, seems v21.12 timeline is suitable for me

Please put this feature complete in v22.04, plan to finish before March

dominicshanshan commented:
> 

found the bug, we need to align two series first, need pandas series API to align two series first. This is have no effect on 'Pearson' correlation but 'Spearman' since it need first align then rank, do we have similar API which I can use right now?
Reference here:

https://github.com/pandas-dev/pandas/blob/v1.4.1/pandas/core/series.py#L2508-L2565
![2022-03-10 at 1 39 PM](https://user-images.githubusercontent.com/30051912/157596910-a2aa023e-a769-4705-a91e-72b62060099b.jpg)

dominicshanshan commented:
> found the bug, we need to align two series first. This is have no effect on 'Pearson' correlation but 'Spearman' since it need first align then rank, do we have similar API which I can use right now?
> 
> https://github.com/pandas-dev/pandas/blob/v1.4.1/pandas/core/series.py#L2508-L2565 ![2022-03-10 at 1 39 PM](https://user-images.githubusercontent.com/30051912/157596910-a2aa023e-a769-4705-a91e-72b62060099b.jpg)

found the corresponding API, _align_indices([self, other], how="inner") will do the job.

dominicshanshan commented:
@isVoid , bug fixed, and already passed pytest in development env, please check

isVoid commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -7478,12 +7478,32 @@ def cov(self, **kwargs):
-        df.columns = self.columns
+        if method == "pearson":
+            corr = cupy.corrcoef(self.values, rowvar=False)
+            df = DataFrame(cupy.asfortranarray(corr)).set_index(self.columns)
```
Is `asfortranarray` necessary here?

isVoid commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -7478,12 +7478,32 @@ def cov(self, **kwargs):
-        df = DataFrame(cupy.asfortranarray(corr)).set_index(self.columns)
-        df.columns = self.columns
+        if method == "pearson":
+            corr = cupy.corrcoef(self.values, rowvar=False)
```
Calling `values` directly may fail if the dataframe is nullable. The test case only captures the situation where the dataframe is non-nullable. Perhaps we should do a dropna here?

isVoid replied:
Out of scope of this PR.

isVoid commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -7478,12 +7478,32 @@ def cov(self, **kwargs):
-        df.columns = self.columns
+        if method == "pearson":
+            corr = cupy.corrcoef(self.values, rowvar=False)
+            df = DataFrame(cupy.asfortranarray(corr)).set_index(self.columns)
```
`.columns` API are external-only API (see https://github.com/rapidsai/cudf/pull/10263). `DataFrame` is not external-only but we should opt for explicit construction whenever possible. Here, I would suggest building the index and columns explicitly and construct the dataframe with `_from_data`.

```suggestion
            column_names = cudf.core.column.as_column(self._column_names)
            index = cudf.core.index._index_from_data({None: column_names})
            df = DataFrame._from_data({col: cudf.core.column.as_column(arr) for col, arr in zip(self._column_names, corr)}, index=index)
```

isVoid replied:
Instead of the suggestion above, you should apply this change. `Dataframe.columns` is a reserved attribute only for external usage due to it's overhead on specific return types to match pandas behavior.
```suggestion
            df = DataFrame(cupy.asfortranarray(corr)).set_index(self._columns_names)
```

isVoid commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -7478,12 +7478,32 @@ def cov(self, **kwargs):
+
+        elif method == "spearman":
+            corr = cupy.corrcoef(self.rank().values, rowvar=False)
+            df = DataFrame(cupy.asfortranarray(corr)).set_index(self.columns)
+            df.columns = self.columns
```
Similar to above.

isVoid replied:
Out of scope of this PR.

isVoid reviewed (changes requested):
Thanks for contributing! I rebased the PR to current development branch 22.04. I have some comments below. The important one is about nullable dataframe handling. You should also fix code formatting issue to trigger CI. For format fixes, see [code-formatting section in contributing guide](https://github.com/rapidsai/cudf/blob/branch-22.04/CONTRIBUTING.md#code-formatting).

And you should also fix code conflicts with upstream.

dominicshanshan commented:
> Thanks for contributing! I rebased the PR to current development branch 22.04. I have some comments below. The important one is about nullable dataframe handling. You should also fix code formatting issue to trigger CI. For format fixes, see [code-formatting section in contributing guide](https://github.com/rapidsai/cudf/blob/branch-22.04/CONTRIBUTING.md#code-formatting).
> 
> And you should also fix code conflicts with upstream.

I will do what you suggested in dataframe, and I saw some changes in v22.04 cudf/dataframe.py (e.g., added 'cols = self._data.to_pandas_index()' in Pearson calculation), so I will alter my dataframe code according this new change and re-check on pytest, for the non-nullable situation I suggest also to add non-nullable in pytest otherwise I do not know if I correctly changed on non-nullable situation? so I will probably re-coding the dataframe.py and you added non-nullable test case in pytest? @isVoid

isVoid reviewed (commented):
> I will do what you suggested in dataframe, and I saw some changes in v22.04 cudf/dataframe.py (e.g., added 'cols = self._data.to_pandas_index()' in Pearson calculation), so I will alter my dataframe code according this new change and re-check on pytest, for the non-nullable situation I suggest also to add non-nullable in pytest otherwise I do not know if I correctly changed on non-nullable situation? so I will probably re-coding the dataframe.py and you added non-nullable test case in pytest? @isVoid

I looked at the code again and I realized some of my suggestions are out of scope of this PR. Supporting nullable dataframe isn't part of today's `corr` API feature. It's best to separate this improvement to a different PR. Please ignore this suggestion.

The suggestion to avoid using `DataFrame` constructor is also out of scope of this PR. So please also ignore that.

I think the only thing that's left is what you mentioned is adding the line of `cols = self._data.to_pandas_index()`, which is a hard constraint because of https://github.com/rapidsai/cudf/pull/10263. So you should insert this line to replace `.columns` usage.

dominicshanshan commented:
> > I will do what you suggested in dataframe, and I saw some changes in v22.04 cudf/dataframe.py (e.g., added 'cols = self._data.to_pandas_index()' in Pearson calculation), so I will alter my dataframe code according this new change and re-check on pytest, for the non-nullable situation I suggest also to add non-nullable in pytest otherwise I do not know if I correctly changed on non-nullable situation? so I will probably re-coding the dataframe.py and you added non-nullable test case in pytest? @isVoid
> 
> I looked at the code again and I realized some of my suggestions are out of scope of this PR. Supporting nullable dataframe isn't part of today's `corr` API feature. It's best to separate this improvement to a different PR. Please ignore this suggestion.
> 
> The suggestion to avoid using `DataFrame` constructor is also out of scope of this PR. So please also ignore that.
> 
> I think the only thing that's left is what you mentioned about adding the line of `cols = self._data.to_pandas_index()`, which is a hard constraint because of #10263. So you should insert this line to replace `.columns` usage. Sorry about the noise.

I will fix the code formatting and close this PR successfully first. 
For non-nullable dataframe, I would love to start a new PR and start to fix it. Since most of hedge fund developer I attached use 'spearman' feature heavily, if is possible I would like to continue coding the correlation part in cuDF.  @isVoid

isVoid commented:
Contributions are always welcome! I created issue https://github.com/rapidsai/cudf/issues/10424 to track correlation for nullable frames. Feel free to take on as you see fit.

dominicshanshan commented:
> @dominicshanshan gentle reminder that release 22.04 burn down starts today. Let us know if you need any help here!

will try to finish today, been busy with other stuff..

dominicshanshan commented:
> > @dominicshanshan gentle reminder that release 22.04 burn down starts today. Let us know if you need any help here!
> 
> will try to finish today, been busy with other stuff..

@isVoid , I passed test_stats.py in pytest both in series.py and dataframe.py. But for formatting issue, please help me, I will learn how you solve this...

dominicshanshan reviewed (commented):
test formatting

isVoid commented:
[This test ](https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fcudf%2Fprb%2Fcudf-style/detail/cudf-style/24582/pipeline) shows `series.py` has format errors.

You can use `black /path/to/series.py` to auto format for you.

dominicshanshan commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -5652,13 +5652,18 @@ def cov(self, **kwargs):
-    @_cudf_nvtx_annotate
-    def corr(self):
+    def corr(self, method="pearson"):
         """Compute the correlation matrix of a DataFrame."""
```
I will take your advice and add back the parameter and return result simple explanation

dominicshanshan reviewed (commented):
black series.py passed

dominicshanshan reviewed (commented):
check again

dominicshanshan commented:
> [This test ](https://gpuci.gpuopenanalytics.com/blue/organizations/jenkins/rapidsai%2Fgpuci%2Fcudf%2Fprb%2Fcudf-style/detail/cudf-style/24582/pipeline) shows `series.py` has format errors.
> 
> You can use `black /path/to/series.py` to auto format for you. After that just create a commit.

@isVoid , I used black to check the style, and it passed black check, but it still failed in build process, could you help me to check?

isVoid commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -5652,20 +5652,34 @@ def cov(self, **kwargs):
+            df.columns = self.columns
+        elif method == "spearman":
+            corr = cupy.corrcoef(self.rank().values, rowvar=False)
+            df = DataFrame(cupy.asfortranarray(corr)).set_index(self.columns)
```
```suggestion
            df = DataFrame(cupy.asfortranarray(corr)).set_index(self._column_names)
```
Similarly.

isVoid commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -5652,20 +5652,34 @@ def cov(self, **kwargs):
+        """
+        if method == "pearson":
+            corr = cupy.corrcoef(self.values, rowvar=False)
+            df = DataFrame(cupy.asfortranarray(corr)).set_index(self._columns_names)
+            df.columns = self.columns
```
```suggestion
            cols = self._data.to_pandas_index()
            df = DataFrame(cupy.asfortranarray(corr)).set_index(cols)
            df._set_column_names_like(self)
```

isVoid commented on python/cudf/cudf/core/dataframe.py:
```diff
@@ -5652,20 +5652,35 @@ def cov(self, **kwargs):
+        DataFrame
+            The requested correlation matrix.
+        """
+        if method == "pearson":
+            corr = cupy.corrcoef(self.values, rowvar=False)
+            cols = self._data.to_pandas_index()
+            df = DataFrame(cupy.asfortranarray(corr)).set_index(cols)
+            df._set_column_names_like(self)
+        elif method == "spearman":
+            corr = cupy.corrcoef(self.rank().values, rowvar=False)
+            df = DataFrame(cupy.asfortranarray(corr)).set_index(self._column_names)
+            df.columns = self.columns
+        else:
+            raise ValueError("method must be either 'pearson', 'spearman'")
```
This part may better be implemented this way to reduce duplication.
```suggestion
        if method == "pearson":
            values = self.values
        elif method == "spearman":
            values = self.rank().values
        else:
            raise ValueError("method must be either 'pearson', 'spearman'")
        corr = cupy.corrcoef(values, rowvar=False)
        cols = self._data.to_pandas_index()
        df = DataFrame(cupy.asfortranarray(corr)).set_index(cols)
        df._set_column_names_like(self)
```

isVoid commented on python/cudf/cudf/core/series.py:
```diff
@@ -2435,9 +2437,18 @@ def corr(self, other, method="pearson", min_periods=None):
-        lhs = self.nans_to_nulls().dropna()
-        rhs = other.nans_to_nulls().dropna()
-        lhs, rhs = _align_indices([lhs, rhs], how="inner")
+        if method == "pearson":
+            lhs = self.nans_to_nulls().dropna()
+            rhs = other.nans_to_nulls().dropna()
+            lhs, rhs = _align_indices([lhs, rhs], how="inner")
+        elif method == "spearman":
+            lhs = self.nans_to_nulls().dropna()
+            rhs = other.nans_to_nulls().dropna()
+            lhs, rhs = _align_indices([lhs, rhs], how="inner")
+            lhs = lhs.rank()
+            rhs = rhs.rank()
+        else:
+            raise ValueError("method must be either 'pearson', 'spearman'")
```
Similarly to reduce duplication:

```suggestion
        lhs = self.nans_to_nulls().dropna()
        rhs = other.nans_to_nulls().dropna()
        lhs, rhs = _align_indices([lhs, rhs], how="inner")
        if method == "spearman":
            lhs = lhs.rank()
            rhs = rhs.rank()
```

isVoid commented on python/cudf/cudf/core/series.py:
```diff
@@ -2422,11 +2422,13 @@ def corr(self, other, method="pearson", min_periods=None):
         """
 
-        if method not in ("pearson",):
+        if method not in ("pearson", "spearman",):
```
```suggestion
        if method not in {"pearson", "spearman"}:
```

isVoid reviewed (changes requested):
I applied some changes to avoid using external only APIs. The below are some new changes to reduce duplication. I think the PR is pretty close!

dominicshanshan commented:
> I applied some changes to avoid using external only APIs. The below are some new changes to reduce duplication. I think the PR is pretty close!

@isVoid , appreciate your help!! why suggest to avoid use external only APIs?

isVoid commented:
> @isVoid , appreciate your help!! why suggest to avoid use external only APIs?

Outside of the context of this PR, there are many use cases where we only want the column names and do some work with it. Internally cuDF stores column names as keys of the [`ColumnAccessor`](https://github.com/rapidsai/cudf/blob/e7dba35d94718c1a7f086af1d14443f0ac17ecb3/python/cudf/cudf/core/column_accessor.py#L81) data structure. Constructing a pandas index often are just to mimic `columns` attribute of a pandas dataframe and is often unnecessary work. [This work](https://github.com/rapidsai/cudf/pull/10263) marks this usage in cuDF apis as an anti-pattern that we will check in tests.

In this PR, however, it's sort of an corner case that we do actually want to explicitly construct an index from the column names. Even so, we would explicitly use the `to_pandas_index` method to construct that so the developer's intent is more explicitly expressed.

For more about `ColumnAccessor`, see: https://docs.rapids.ai/api/cudf/stable/basics/internals.html#columnaccessor

isVoid commented on python/cudf/cudf/core/series.py:
```diff
@@ -2422,11 +2422,13 @@ def corr(self, other, method="pearson", min_periods=None):
         >>> ser1 = cudf.Series([0.9, 0.13, 0.62])
         >>> ser2 = cudf.Series([0.12, 0.26, 0.51])
-        >>> ser1.corr(ser2)
+        >>> ser1.corr(ser2, method="pearson")
```
I think `Series.corr` can add a "parameter" section explaining `method`.

isVoid reviewed (approved):
Thanks! Just one more comment that you can feel free to address at your own discretion.

brandon-b-miller reviewed (approved):
LGTM! great job and thanks!

bdice commented:
@gpucibot merge
